### PR TITLE
Helper Function for processing URLS

### LINKED ISSUE #5: Add YT-related functionality
- [x] retrieve top 5 videos based on query
- [x] recognise yt video links
- [x] recognise yt playlist links and retrieve the videos within

### PR DESCRIPTION
**Pull request by raresboza**
Example: ``` https://www.youtube.com/playlist?list=PLtsnbIyoYa4R0ynUgJtoMmY20L3KqTq5L``` becomes
```
[
  { source: 'YouTube', id: 'X_DxYliRWnU' },
  { source: 'YouTube', id: 'JKPp5rWcElw' },
  { source: 'YouTube', id: '0PS1g_aiUcs' },
  { source: 'YouTube', id: 'nRzoCIALsBg' },
  { source: 'YouTube', id: '74gnB_7eFlE' },
  { source: 'YouTube', id: 'ZgbJb6ybNe4' },
  { source: 'YouTube', id: 'bzxz2j6NDVc' },
  { source: 'YouTube', id: 'ZJXXIOnlzpg' },
  { source: 'YouTube', id: 'jBzPrhsQWCg' },
  { source: 'YouTube', id: '4uLCXJGIdhs' },
  { source: 'YouTube', id: 'Ny5y7USoQoA' },
  { source: 'YouTube', id: 'JvJHoqpdGQc' },
  { source: 'YouTube', id: 'JhxLqard9z0' }
]
 
```

### CODE REVIEW

File: utils/link_utils.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    return { source: null, id: null };
+}
+
+// converts yt playlist to a list of (YouTube, <url>)
+async function getYtItemsFromPlaylist(playlistId) {
+    const url = `https://www.googleapis.com/youtube/v3/playlistItems?part=snippet&playlistId=${playlistId}&maxResults=100&key=${ytApiId}`;
+    try {
+        const response = await axios.get(url);
+        const videoIds = response.data.items.map(item => ({ source: 'YouTube', id: item.snippet.resourceId.videoId }));
```
Comment (raresboza) [inline utils/link_utils.js]: currently I hardcode source to the same string as the key in the dictionary with with regex.
I am not sure if we should use numbers (less clarity, less memory usage), or if there is another way I can handle this.

(Context: we need to change a YouTube playlist to a list of YouTube links. Given that we might also offer support for other platforms, we need to tag the links appropriately).

File: utils/link_utils.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    catch (error) {
+        console.error('Error fetching playlist items:', error);
+        return [];
+    }
+}
+
+async function processUrl(url) {
+    const { source, id } = identifyUrlType(url);
+
+    if (source === 'YouTubePlaylist') {
```
Comment (raresboza) [inline utils/link_utils.js]: same as above

File: utils/link_utils.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    YouTubePlaylist: /^(?:https?:\/\/)?(?:www\.)?youtube\.com\/playlist\?list=([a-zA-Z0-9_-]+)/,
+};
+
+// matches url against regex
+function identifyUrlType(url) {
+    for (const source in sourceRegexList) {
+        const regex = sourceRegexList[source];
+        const match = url.match(regex);
+        if (match) {
+            return { source, id: match[1] };
```
Comment (giobiba) [inline utils/link_utils.js]: can we have here return { source: source, id: match[1] };  to keep the standard from line 19

File: utils/link_utils.js (hunk 2 above)
Comment (giobiba) [reply utils/link_utils.js]: we could use enumerations for source type

File: utils/link_utils.js (hunk 1 above)
Comment (giobiba) [reply utils/link_utils.js]: same as below

**Review by giobiba — COMMENTED** [review]
some thoughts :)

**Comment by raresboza** [comment]
Closed, feedback applied

**Pull request merged**